### PR TITLE
Perma Payments integration

### PR DESCRIPTION
**Pull request by rebeccacremona**
This is less polished than the Perma Payments app, but I hope it will serve.

The plot
---------
There are two kinds of registrars: those who qualify for a free account, and those who do not. By default, all new registrars are "nonpaying".  To indicate that a registrar does not qualify for a free account, just view that registrar in the Perma admin, toggle the "nonpaying" field, specify their monthly rate, and save.

As soon as that is done, the features below are enabled.

### Settings -> Subscription
All registrar users for the registrar have access to a new settings page, Settings -> Subscription, where they may purchase a Perma.cc subscription (via CyberSource via Perma Payments). After they purchase a subscription, Settings -> Subscription provides them with information about their subscription: status, next payment date, rate, etc., and gives them the opportunity to update their payment information (again, via CyberSource via Perma Payments), or to cancel their subscription.

### Making Links
Every time a user associated with that registrar attempts to make a Permalink in a folder associated with the registrar, rather than their Personal Links folder, Perma.cc calls out to Perma Payments to find out if the registrar has a paid-up, current subscription. If yes, capture proceeds. If not, the user is informed that Perma.cc currently can't make Permalinks for their registrar, and is given appropriate information to resolve: org users are put in touch with their registrar users; registrar users are directed to Settings -> Subscription. 

If that communication fails for any reason, it's no problem: Perma.cc keeps a record of the last known subscription status for each registrar, and will use that value to make its determination. If the communication fails and Perma doesn't have any known statuses on record, the link is not created.

### Everything Else
Everything else is handled by Perma Payments.

Technical Details
-----------------

### Config
Some new configuration variables are required, all of which are secret or semi-secret and all of which relate to communication with Perma Payments. No defaults are provided in settings_common.py, so that the app will complain if you are missing something you need. See settings.example.py for required keys.

### New Fixtures
There are several new dev fixtures available: a Test Firm, one registrar user, and two org users. See users.json for their email addresses. Their passwords are all the usual value.

I recommend using fab dev.reset_hard_db to throw away your existing database, and then making a new one with init_db. Any warcs you had will still be in your repo's services directory, if you want them, but you won't see any links via the perma.dev:8000 interface.

### Communicating with the outside world
Perma.cc calls out to Perma Payments for information and directs users to Perma Payments for all subscription-related transactions. It's unclear how your local Perma, inside the vagrant box, should behave.

Option 1: If you have enough RAM, you can spin up a local Perma Payments using Docker, and finesse things so that your local Perma and Perma Payments can speak to each other. Since local Perma runs on port 8000 by default and local Perma Payments runs on port 80 by default, it's not as challenging as it might sound. It can be convenient, especially if you are working on the interface for the two apps.

Option 2: You can configure your Perma to talk to the stage tier of Perma Payments, which is live on AWS. Ask RLC for the config, if you like. This is the fastest to get working. It's reasonably convenient, though, since we are all working with the same Perma dev fixtures, we could accidentally clobber each other (if one person wants registrar 4 to be in status "On Hold" and another person sets it to "Current", etc.)

Option 3: (hack) You can put a short in the Perma Code to get the result you want. E.g., throw `return True` or `return False` in at the top of `models.Registrar.link_creation_allowed`, or temporarily arrange for `models.Registrar.get_subscription` to return a dict of your liking rather than call out to Perma Payments. Just don't commit your changes. (If you do, tests will fail, so don't worry too much. I actually did this several times......)

If anybody has better ideas, let me know. I thought about putting a config variable in.... but I'm not sure I like that either....

Testing mocks out the python requests library, so no real network calls are made.

For Review
------------
Hopefully this will be MUCH less work than reviewing Perma Payments. There's not a whole lot going on. Any time you have, I appreciate, but if you are too busy, I don't think there's a lot of risk here.

### Views
Do you see anything too rough to go live with?

### Code
Do you see anything inscrutable, that needs a comment (or a refactor!)?
Do you see anything obviously wrong?
Do you see anything than can obviously be done in a cleaner/better/more idiomatic way?

The code for communicating with Perma Payments is the same as the code in Perma Payments, for communicating with Perma, except back-ported to python 2.

The tests are pretty poor, but I'm hoping they'll do: feel free to ignore.

### CODE REVIEW

**Comment by rebeccacremona** [comment]
image
image
image
image
image
image

Subscription not paid up, if you are a registrar user:
image

Subscription not paid up, if you are an org user:
image

File: perma_web/perma/models.py [hunk 1]
```diff
@@ -169,6 +201,99 @@ def most_active_org_this_year(self):
...
+        r = requests.post(
+            settings.SUBSCRIPTION_STATUS_URL,
+            data={
+                'encrypted_data': prep_for_perma_payments({
+                    'timestamp': to_timestamp(datetime.utcnow()),
+                    'registrar':  self.pk
+                })
+            }
+        )
+        if r.status_code != 200:
```
Comment (anastasia) [inline perma_web/perma/models.py]: wondering if status code can ever be 2xx success but not 200?

File: perma_web/perma/models.py [hunk 2]
```diff
@@ -446,6 +571,11 @@ def is_supported_by_registrar(self):
         return settings.CONTACT_REGISTRARS and \
                self.is_organization_user
 
+    def can_view_subscription(self):
+        registrar = self.registrar
+        return registrar and not registrar.nonpaying
```
[the remark this answers is not in the document]
Comment (rebeccacremona) [reply perma_web/perma/models.py]: Ah, totally.

File: perma_web/perma/templatetags/pretty_date.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import inflect
+
+from django import template
+
+register = template.Library()
+
+@register.filter
+def pretty_date(date):
```
[the remark this answers is not in the document]
Comment (rebeccacremona) [reply perma_web/perma/templatetags/pretty_date.py]: Aha. I..... have no idea. I'll find out :-)

File: perma_web/perma/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rebeccacremona) [reply perma_web/perma/models.py]: Interesting. I was using asserts for flow in Perma Payments, and removed them after a security linting library complained: evidently, if somebody runs python with the -o flag (for 'optimized'), all asserts are stripped out. But we, of course, are not running with the -o flag. I like asserts! Will do this here.

@anastasia also remarked on !=200 being too specific. response.ok sounds just right.

[the remark this answers is not in the document]
Comment (jcushman) [reply perma_web/perma/models.py]: Yeah, -o flag feels pretty hypothetical, but I suppose you could do try-if-raise-except ...

```
try:
    response = requests.get()
    if not response.ok:
        raise AssertionError("daaaaang")
except (requests.RequestException, AssertionError) as e:
    raise PermaPaymentsCommunicationException(str(e))
```

File: perma_web/perma/templatetags/pretty_date.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rebeccacremona) [reply perma_web/perma/templatetags/pretty_date.py]: Yep. How about that.

**Pull request merged**